### ISSUE TITLE
RandomForests are very slow

### ISSUE THREAD

**Issue by denisld** [mlpack/mlpack#3503]
Hello,

I am currently evaluating a number of C++-based RandomForest implementations, as an alternative to the Python-based scikit-sklearn tools. To my surprise, C++ MLPack is much, much slower than the Python-based SKLearn implementation. I hope somebody can help me understand the reasons for this.

Some details:

- I have a dataset consisting of 21 million datapoints. There are 7 features per datapoint, and 2 labels.
- I am using 70% of this dataset for training, 30% for validation.
- I am training a random forest consisting of 150 trees, with a maximum depth of 50 and a minimum node size of 1.
- I have done the above both in C++/MLPack and Python/scikit-learn. 
- In the MLPack version I used my own top-level main(), so I could enable OpenMP support. I also briefly used the pre-built command-line version, but it was slower than my own version, because it didn't use multithreading/OpenMP.
- The MLPack version is built on a Mac with OpenMP support (for Armadillo), as a native arm64 target with clang, using -O3 for optimization. Activity Monitor shows that MLPack is indeed using 10-ish cores for training, and approx 3 cores for classification.

Here are the results:

- Both C++/MLPack and Python/scikit-learn reach a similar accuracy (91.6% for MLPack, 92.6% for scikit).
- The C++/MLPack version takes approximately **4000 seconds** to train. The Python/scikit-learn version takes **1500 seconds**. I consider this a small but significant advantage for Python/scikit. 
- To classify the 30% validation points, C++/MLPack takes **more than 5 hours**, whereas Python/scikit-learn does this in **93 seconds**. This is the bigger problem.

I have studied the source code of both scikit-learn and MLPack to some extent, and I cannot immediately find an explanation for this massive performance difference, on what should pretty much be the same algorithm. If anything, we should expect the C++ version to be somewhat faster, not 200 times slower.

I have found a similar issue, but that one was automatically closed without a proper resolution: https://github.com/mlpack/mlpack/issues/1643 

I'd appreciate any insight on this. I hope one of the devs could reproduce my results by running a similar simple head-to-head test in python/scikit-learn and MLPack. Thanks in advance for any help!

**Comment by rcurtin** [comment]
Thanks for the detailed report.  I'm surprised by what you are finding.  I'm interested to know whether the mlpack trees are just significantly larger than the scikit trees... do you think you could collect this statistic?

It's a little tedious but here is some code that can count the number of nodes in a random forest:

```
RandomForest<...> r; // train however ...

size_t count = 0;
for (size_t i = 0; i < r.NumTrees(); ++i)
{
  DecisionTree<...>& d = r.Tree(i);
  std::queue<DecisionTree<...>*> q;
  q.push(&d);
  while (!q.empty())
  {
    ++count;
    DecisionTree<...>* n = q.front();
    q.pop();
    for (size_t j = 0; j < n->NumChildren(); ++j)
      q.push(&n->Child(j));
  }
}
std::cout << "Number of nodes in random forest: " << count << ".\n";
```

You'll need to fix the template parameters depending on how you configured the random forest, but that should work.  If not, let me know, it is probably a simple error (I did not compile and test what I just wrote, but I think it's right...).

If that doesn't give an obvious smoking gun, is there any way you could share the dataset?  If not, the dataset size you gave looks a lot like the higgs dataset, which can be found on the UCI datasets repository or here: https://datasets.mlpack.org/higgs.csv.gz and https://datasets.mlpack.org/higgs.labels.csv.gz.  Can you see if that dataset gives the same behavior?  If so, I can reproduce that locally and try to figure out what the holdup is here.  5 hours to classify 6.3M points is an average of 350 seconds per point!  Something is definitely wrong.

**Comment by denisld** [comment]
Thanks so much for the quick reply! I will run the code now to get that metric. It will take a while, obviously.

By the way, rather than figuring out the template parameters by hand, it's easier to use auto and decltype:

`  size_t count = 0;
    for (size_t i = 0; i < forest.NumTrees(); ++i)
    {
        auto& d = forest.Tree(i);
        std::queue<std::remove_reference<decltype(d)>::type*> q;
        q.push(&d);
        while (!q.empty())
        {
            ++count;
            auto* n = q.front();
            q.pop();
            for (size_t j = 0; j < n->NumChildren(); ++j)
                q.push(&n->Child(j));
        }
    }
    std::cout << "Number of nodes in random forest: " << count << ".\n";
`
I am able to share the original data set with you, but it is 1.21 Gb when compressed. The attachment limit for comments is 25 MB.

**Comment by rcurtin** [comment]
Yeah, I thought about using `auto`, but it was too late by the time I wrote the comment. :)  If you have another way to share the dataset, that works; maybe upload to something like Google Files or Dropbox and send me a link?  You can keep the link private if you like and email it to me--- ryan@ratml.org. :+1:

**Comment by denisld** [comment]
Thanks again! I sent you a link with the dataset, and I started a new run to count the nodes. Will let you know when I have that info.

**Comment by rcurtin** [comment]
Using the Higgs dataset I was not able to reproduce the prediction time slowness.  Here's the transcript of using the command-line `mlpack_random_forest` to train a model, and then predict on the full 11M-point dataset:

```
$ OMP_NUM_THREADS=8 bin/mlpack_random_forest -t /archive/datasets/higgs.csv -l /archive/datasets/higgs.labels.csv -T /archive/datasets/higgs.csv -v --num_trees 150
[INFO ] Loading '/archive/datasets/higgs.csv' as CSV data.  Size is 29 x 11000000.
[INFO ] Loading '/archive/datasets/higgs.labels.csv' as raw ASCII formatted data.  Size is 11000000 x 1.
[INFO ] Training random forest with 150 trees...
[INFO ] Loading '/archive/datasets/higgs.csv' as CSV data.  Size is 29 x 11000000.
[INFO ]
[INFO ] Execution parameters:
[INFO ]   help: 0
[INFO ]   info:
[INFO ]   input_model_file:
[INFO ]   labels_file: '/archive/datasets/higgs.labels.csv' (11000000x1 matrix)
[INFO ]   maximum_depth: 0
[INFO ]   minimum_gain_split: 0
[INFO ]   minimum_leaf_size: 1
[INFO ]   num_trees: 150
[INFO ]   output_model_file:
[INFO ]   predictions_file: ''
[INFO ]   print_training_accuracy: 0
[INFO ]   probabilities_file: ''
[INFO ]   seed: 0
[INFO ]   subspace_dim: 0
[INFO ]   test_file: '/archive/datasets/higgs.csv' (11000000x29 matrix)
[INFO ]   test_labels_file: ''
[INFO ]   training_file: '/archive/datasets/higgs.csv' (11000000x29 matrix)
[INFO ]   verbose: 1
[INFO ]   version: 0
[INFO ]   warm_start: 0
[INFO ] Program timers:
[INFO ]   loading_data: 164.974739s (2 mins, 44.9 secs)
[INFO ]   rf_prediction: 84.411163s (1 mins, 24.4 secs)
[INFO ]   rf_training: 1083.254256s (18 mins, 3.2 secs)
[INFO ]   total_time: 1217.608973s (20 mins, 17.6 secs)
```

84 seconds is far more in line with what I would have expected; so something must be different about the dataset you're using.  Anyway, I am working on reproducing with the dataset you sent, so we'll see what that does.

**Comment by denisld** [comment]
Ok, interestingly, I am not able to re-run my own dataset, because the application keeps getting killed by the kernel. Logs show this is due to an out-of-memory error, and I think there is a clue here that might explain the performance problem.

Some more details about my system that may be relevant: I'm running on Apple M1 Max silicon, compiling native code, I have 32 GB of physical RAM, and (obviously, I suppose) an SSD.

If I look at the RAM usage during the training phase, it shoots up fairly quickly (currently around 60 GB, after about 15 minutes). This could mean that my system is disk-thrashing, which could account for a significant performance loss. I am going to try to re-run the experiment on a Linux machine with much more memory this afternoon, and report back.

**Comment by denisld** [comment]
Hi again. Over the past week I have completed a series of measurements comparing the performance of random forests in MLPack (C++) versus Scikit learn (Python), measuring both runtime and peak memory usage. So far I've only focused on training, but I am tentatively concluding the following:

- MLPack/C++ uses five to six times as much memory as Scikit/Python during training.
- MLPack/C++ uses a different, significantly slower, algorithm than Python/Scikit learn to train the forest.

I'll present the details to support these claims here.

### Test Setup

- I have used a core dataset (as described in an earlier post) consisting of 21 million points, 7 features, an 2 labels.
- For the tests I've used randomly chosen subsets of the core dataset of increasing sizes. Up until 2 million points I have used increments of 100k points between runs, from 2 million onward I have used increments of 1 million points, up to and including 11 million points in total. I stopped after that, because MLPack was reaching the limits of physical RAM on my test machine.
- I have trained random forests in both Scikit/Python and MLPack/C++ of 150 trees, max depth 50, minimum node size 1. 
- I have performed all measurements on an M1 MacBook Pro with 32 GB of physical RAM.
- I compiled the MLPack-application with clang, OpenMP support enabled, and -O3 optimization.
- For verification, I also ran some of the measurements on a Linux-based Core i5 system with 128 GB of physical RAM. On that system I compiled using gcc, also with OpenMP and -O3. The results of these verification runs are not listed below, but they show a similar development as on the Mac.
- I have compared training accuracies of runs between both implementations. They are consistently similar (around 90% accuracy for both implementations).

### Measurement Results

The following table lists the raw measurement data for Scikit Learn. Columns are: **size of the training data set** in points, **real time** (actual elapsed time), **user time** (CPU time spent in the application, which due to the number of cores, is some multiple of the real time), **system time** (time spent in the kernel during training), and **peak memory usage** in bytes.

Performance data for Scikit-learn/Python:
```
100000 4.81 38.7 2.85 302451968
200000 10.11 85.86 3.1 502714560
300000 15.58 135.83 3.19 400279936
400000 22.24 188.26 3.32 781015040
500000 28.6 250.66 3.35 800135488
600000 34.25 296.65 3.93 715934976
700000 42.43 376.48 4.36 839814400
800000 50.25 449.12 4.25 964765952
900000 56.41 498.63 4.55 1061038144
1000000 62.22 554.83 4.86 1230269120
1100000 71.04 636.1 5.41 1746151104
1200000 79.94 712.69 5.84 1877845760
1300000 85.65 760.35 6.53 2089953344
1400000 99.18 896.88 6.34 1564070464
1500000 99.66 890.7 6.73 1677022080
1600000 106.7 959.32 6.92 1774883776
1700000 118.42 1062.03 7.02 1894569344
1800000 126.74 1140.78 7.39 2029191488
1900000 131.93 1189.81 7.19 2133737408
2000000 140.74 1266.91 7.59 2264580352
3000000 223.34 2010.43 10.36 3256721536
4000000 321.7 2779.42 18.9 4329613504
5000000 403.27 3481.81 20.49 5366460608
6000000 487.3 4290.96 18.6 6411810560
7000000 584.9 5161.42 23.18 7471185984
8000000 659.41 5819.0 29.66 8523171584
9000000 770.94 6739.4 32.21 9590476928
10000000 836.85 7343.95 36.61 10642855872
11000000 930.94 8104.61 34.92 11671953088
```

Next is the same table, but for MLPack/C++:

```
100000 2.64 16.19 0.69 753240576
200000 5.75 37.47 1.67 1478214528
300000 9.55 60.37 1.93 2150328320
400000 13.65 86.41 2.69 2857276672
500000 16.84 115.67 4.1 3519806208
600000 26.62 146.35 6.95 4154300096
700000 26.02 174.31 6.21 4799312128
800000 32.07 214.61 7.32 5480600128
900000 36.64 258.44 7.89 6088040640
1000000 41.25 301.04 9.48 6715259968
1100000 47.18 356.13 10.34 7447343296
1200000 53.62 395.52 10.88 8021732992
1300000 58.69 437.01 12.67 8612394240
2000000 109.97 854.06 24.46 12984639616
3000000 207.81 1620.76 41.59 18922882240
4000000 319.29 2495.41 63.93 24809241152
5000000 470.06 3640.7 79.76 30741774848
6000000 628.64 4858.35 106.18 36105981824
7000000 801.75 6126.89 127.53 41713928320
8000000 964.94 7359.92 138.65 47050227392
9000000 1220.55 9237.84 162.8 53564715520
10000000 1418.23 10655.33 183.12 59797078784
11000000 1541.55 11596.74 199.77 72848490176
```

### Discussion

Below is a plot of the (real) **time** of both MLPack/C++ and Scikit/Python:

<img width="2592" alt="MLPackVSScikitRuntime" src="https://github.com/mlpack/mlpack/assets/16610460/1b5dcd83-3d61-4ef5-aafd-cf30c1f67041">

This plot strongly suggests that the training time in Scikit scales linearly with the number of points, whereas MLPack training time scales according to some higher order function. Around 4 million points, the inherent speed benefit of C++ over Python starts to lose to the apparent higher time complexity of MLPack.

The following plots show the **peak memory usage** of MLPack/C++ and Scikit/Python as a function of dataset size:

<img width="2592" alt="MLPackVSScikitPeakMemoryUsage" src="https://github.com/mlpack/mlpack/assets/16610460/b4a7487a-bccb-47ed-a514-46c2bf83c27d">

It appears to show that the space complexity of both implementations is linear, with MLPack having a significantly larger memory footprint. 

The **ratio** between memory size of MLPack and  as a function of dataset size is as follows:

<img width="2592" alt="MLPackVSScikitPeakMemoryRatio" src="https://github.com/mlpack/mlpack/assets/16610460/6938e7b1-3de3-48b4-9e33-29057b5bbc6f">

...which leads me to conclude that MLPack uses upward of 5.5 times as much memory as Scikit.

### Conclusion

I think I've been able to show that there is a structural difference in training performance between the C++-based MLPack and the Python-based Scikit Learn. It appears that the difference is due to a fundamental algorithmic difference between the two implementations. It might be that MLPack consciously uses a different training algorithm than Scikit, or it could be an unintended effect of inappropriate data structure choices. 

N.B., I'd like to repeat that so far my experiments have focused on the performance of the training process. I am in the process of performing classification performance measurements as well.

I look forward to your insights on this. Thanks in advance.

**Comment by conradsnicta** [comment]
@denisld Thanks for the analysis. mlpack is a volunteer driven project, and many people are working on various projects. It may take a while before we can properly look at this.

**Comment by rcurtin** [comment]
To provide an update on this: I believe I have isolated an issue, but I haven't figured out the fix yet.  I am traveling right now, so time is difficult, but it's basically at the top of my todo list once I get home next week (if I don't manage to get it by then).

**Comment by rcurtin** [comment]
So it has taken a while to get back to this due to travel, etc., but the short story of what I discovered is that there was a bug with RNG seeding, which meant (roughly) that every tree trained in a random forest when using OpenMP was exactly the same.  (That is exactly true if the number of trees is less than or equal to the number of threads.)  More details in #3521.

I then ran my own scaling experiments that were slightly different than yours.  I was interested in the accuracy issue---mlpack should be able to produce forests that are as accurate as scikit's.  So I used the dataset you provided and swept the number of trees, tracking the training time, test time, and accuracy on the training and test sets.  I used the `mlpack_random_forest` command-line program, which should be pretty equivalent to what you were doing.  I didn't make a pretty graph but I did make a pretty text file:

```
num trees     train time    test time     train acc     test acc

1              94.772s        2.354s      94.3082%      86.3599%
2              97.199s        4.557s      95.2071%      87.6581%
3              94.874s        7.012s      97.5387%      89.6952%
4              98.971s        9.732s      97.8425%      90.1066%
5             101.216s       12.528s      98.5236%      90.7431%
10            106.726s       28.359s      99.3226%      91.6845%
15            126.113s       45.298s      99.6169%      92.0726%
20            144.901s       68.112s      99.7465%      92.2352%
25            227.586s       82.931s      99.8217%      92.3506%
30            238.254s      102.723s      99.8678%      92.4108%
35            242.672s      123.731s      99.8977%      92.4648%
40            263.150s      143.752s      99.919%       92.4873%
45            286.809s      165.133s      99.9342%      92.5167%
50            362.554s      189.399s      99.9461%      92.5384%
60            382.941s      228.890s      99.9618%      92.5732%
70            426.322s      284.815s      99.9721%      92.5899%
80            519.825s      329.458s      99.9785%      92.6066%
90            544.712s      368.747s      99.983%       92.6179%
100           652.263s      429.334s      99.9867%      92.6322%
110           666.127s      462.426s      99.9893%      92.6344%
120           702.539s      524.086s      99.9914%      92.6402%
130           790.670s      562.567s      99.9933%      92.6466%
140           825.846s      667.973s      99.9942%      92.6493%
150           917.266s      707.518s      99.9951%      92.6529%
```

For those simulations, I did not set `--max_depth` (i.e. the trees could be however deep).  So, when the settings are basically the same as scikit in your experiments, with the changes in #3521, the accuracy is now 92.65%, which may even be slightly better (but certainly not much).  Test time appears to scale roughly linearly with the number of trees.

I would be really interested to see if (or how much) this fixes the original issue you were reporting.  Is it possible to re-run your experiments with the changes in #3521?  I'd love to see if the number of nodes in each forest has changed.  (I am not sure it will, but that at least gives me a bit more information to keep debugging if needed.)

Thanks again for the report!

**Comment by conradsnicta** [comment]
@denisld #3521 was merged.  Let us know if this addresses the issue on your setup.

**Comment by denisld** [comment]
Thanks for your effort so far. I'll try to re-run part of the experiment in the coming days, and I'll let you know if the change has any effect on the runtime.
